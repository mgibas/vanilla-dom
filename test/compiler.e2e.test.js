const compiler = require('../src/compiler')

describe('compiler', () => {
  let options
  beforeEach(()=>{
    options = {state:'st', module: 'closure'}
    window.document.body.innerHTML = ''
  })

  describe('mounting compiled code', () => {
    it.each([
      ['empty', ''], 
      ['simple', '<div></div>'], 
      ['with attributes', '<div foo="5" bar="some"></div>'], 
      ['multiple nodes', '<div></div><p></p>'], 
      ['text nodes', ` <div></div>some text and empty line
        
        <p></p>`], 
      ['nested', `<div><p>hello</p></div>`]
    ])('mount function recreates given template on a given node - %s',
      (testcaseName, template) => {
        let compiled = compiler.compile(template, options)
        let mount = eval(compiled)
        mount(window.document.body)
        expect(window.document.body.innerHTML).toMatchSnapshot();
      }
    );

    it.each([
      ['simple expressions', '<div foo="{{st.foo}}">{{st.fullname}}</div>'],
      ['property expression on undefined property', '<div foo="{{st.missing}}">{{st.missing}}</div>'],
      ['boolean expression true', '<div bar="{{!st.zero}}">{{!st.zero}}</div>'],
      ['boolean expression false', '<div bar="{{!!st.zero}}">{{!!st.zero}}</div>'],
      ['property expression with 0 as result', '<div foo="{{st.zero}}">{{st.zero}}</div>'],
      ['property expression with empty string as result', '<div bar="{{st.empty}}">{{st.empty}}</div>'],
      ['nested property expression on undefined property', '<div foo="{{st.missing.missingbad}}">{{st.missing.missingbad}}</div>'],
      ['multiple expressions with undefined properties', '<div foo="{{st.missing}}">{{st.missing}}  and {{st.somissing}}</div>'],
      ['multiple expressions', '<div foo="{{st.foo}} and {{st.bar}}">{{st.foo}} and {{st.bar}}</div>'],
      ['javascript expressions', '<div sum="{{st.foo + st.bar + 2}}">{{st.foo + st.bar}}</div>'],
      ['method expressions', '<div upper="{{st.fullname.toUpperCase()}}">{{st.firstname.toUpperCase()}}</div>']
    ])('mount function evaluates given expression - %s',
      (testcaseName, template) => {
        let state = {
          fullname: 'Maciej Gibas',
          firstname: 'Maciej',
          foo: 2,
          bar: 3,
          zero: 0,
          empty: ''
        }
        let compiled = compiler.compile(template, options)
        let mount = eval(compiled)
        mount(window.document.body, state)
        expect(window.document.body.innerHTML).toMatchSnapshot();
      }
    );

    it.each([
      ['simple expressions', '<div foo="{{st.foo}}">{{st.fullname}}</div>'],
      ['property expression on undefined property', '<div foo="{{st.missing}}">{{st.missing}}</div>'],
      ['nested property expression on undefined property', '<div foo="{{st.missing.missingbad}}">{{st.missing.missingbad}}</div>'],
      ['boolean expression from true to false', '<div bar="{{!!st.nonZero}}">{{!!st.nonZero}}</div>'],
      ['boolean expression from false to true', '<div bar="{{!!st.zero}}">{{!!st.zero}}</div>'],
      ['property expression from 0 value to non 0', '<div foo="{{st.zero}}">{{st.zero}}</div>'],
      ['property expression from non 0 to 0 value', '<div foo="{{st.nonZero}}">{{st.nonZero}}</div>'],
      ['boolean expression from empty to non empty', '<div bar="{{st.empty}}">{{st.empty}}</div>'],
      ['boolean expression from non empty to empty', '<div bar="{{st.nonEmpty}}">{{st.nonEmpty}}</div>'],
      ['multiple expressions with undefined properties', '<div foo="{{st.missing}}">{{st.missing}}  and {{st.somissing}}</div>'],
      ['multiple expressions', '<div foo="{{st.foo}} and {{st.bar}}">{{st.foo}} and {{st.bar}}</div>'],
      ['javascript expressions', '<div sum="{{st.foo + st.bar + 2}}">{{st.foo + st.bar}}</div>'],
      ['method expressions', '<div upper="{{st.fullname.toUpperCase()}}">{{st.firstname.toUpperCase()}}</div>']
    ])('update state in dom on update call - %s',
      (testcaseName, template) => {
        let state = {
          fullname: 'Maciej Gibas',
          firstname: 'Maciej',
          foo: 2,
          bar: 3,
          nonZero: 1,
          zero: 0,
          empty: '',
          nonEmpty: 'blah'
        }
        let updatedState = {
          fullname: 'Foo Bar',
          firstname: 'Foo',
          foo: 300,
          bar: 500,
          nonZero: 0,
          zero: 1,
          empty: 'blah',
          nonEmpty: ''
        }
        let compiled = compiler.compile(template, options)
        let mount = eval(compiled)
        let update = mount(window.document.body, state)
        update(updatedState)
        expect(window.document.body.innerHTML).toMatchSnapshot();
      }
    );

    describe('arrays', () => {
      it.each([
        ['simple', '<ul><li repeat-for="{{st.objects}}">{{items[i].name}}</li></ul>'],
        ['with attributes', '<ul><li repeat-for="{{st.objects}}"><p data-name="{{items[i].name}}">{{items[i].name}}</p></li></ul>'],
        ['custom "as"', '<ul><li repeat-for="{{st.objects}}" repeat-as="rows">{{rows[i].name}}</li></ul>'],
        ['custom "index"', '<ul><li repeat-for="{{st.objects}}" repeat-index="b">{{items[b].name}}</li></ul>'],
        ['js expression', '<ul><li repeat-for="{{st.objects.concat(st.objects)}}">{{items[i].name}}</li></ul>'],
        ['access index', '<ul><li repeat-for="{{st.objects}}">index: {{i}}</li></ul>'],
        ['nested template', '<ul><li repeat-for="{{st.objects}}"><p>some text</p><div><span>more</span></div></li></ul>'],
        ['with sybling', '<div><span repeat-for="{{st.objects}}">{{i}}</span><div>I like this place</div></div>']//,
        //['nested array', `
          //<ul>
            //<li repeat-for="{{st.objects}}">
              //<span repeat-for="{{items[i].chars}}" repeat-as="chars" repeat-index="j">{{chars[j]}}</span>
            //</li>
          //</ul>`]
      ])('mounts array using initial state data - %s', (testcaseName, template) => {
        let state = {objects: [{name: 'foo', chars: ['f','o','o']},{name: 'bar', chars: ['b','a','r']}]}
        let compiled = compiler.compile(template, options)
        let mount = eval(compiled)
        mount(window.document.body, state)
        expect(window.document.body.innerHTML).toMatchSnapshot();
      })

      it.each([
        ['simple', '<ul><li repeat-for="{{st.objects}}">{{items[i].name}}</li></ul>'],
        ['custom "as"', '<ul><li repeat-for="{{st.objects}}" repeat-as="rows">{{rows[i].name}}</li></ul>'],
        ['custom "index"', '<ul><li repeat-for="{{st.objects}}" repeat-index="b">{{items[b].name}}</li></ul>'],
        ['js expression', '<ul><li repeat-for="{{st.objects.concat(st.objects)}}">{{items[i].name}}</li></ul>']//,
        //['nested array', `
          //<ul>
            //<li repeat-for="{{st.objects}}">
              //<span repeat-for="{{items[i].chars}}" repeat-as="chars" repeat-index="j">{{chars[j]}}</span>
            //</li>
          //</ul>`]
      ])('update without array length change - %s', (testcaseName, template) => {
        let state = {objects: [{name: 'foo', chars: 'foo'.split('')},{name: 'bar', chars: 'bar'.split('')}]}
        let newState = {objects: [{name: 'foo-updated', chars: 'foo-updated'.split('')},{name: 'bar-updated', chars: 'bar-updated'.split('')}]}
        let compiled = compiler.compile(template, options)
        let mount = eval(compiled)
        let update = mount(window.document.body, state)

        update(newState)

        expect(window.document.body.innerHTML).toMatchSnapshot();
      })

      it.each([
        ['add item', [{name: 'foo'},{name: 'bar'}], [{name: 'foo'},{name: 'bar'},{name: 'oof'}]],
        ['no array to array with items', null, [{name: 'foo'},{name: 'bar'}]],
        ['remove item', [{name: 'foo'},{name: 'bar'},{name: 'oof'}], [{name: 'foo'},{name: 'oof'}]]
      ])('splicing - %s', (testcaseName, from, to) => {
        let state = {objects: from}
        let newState = {objects: to}
        let compiled = compiler.compile('<ul><li repeat-for="{{st.objects}}">{{items[i].name}}</li></ul>', options)
        let mount = eval(compiled)
        let update = mount(window.document.body, state)

        update(newState)

        expect(window.document.body.innerHTML).toMatchSnapshot();
      })
    })
    describe('events handlers', () => {
      it('simple handler', () => {
        let ctx = {fire: jest.fn()}
        let template = `<button on-click="this.fire" />`
        let compiled = compiler.compile(template, options)
        let mount = eval(compiled).bind(ctx)
        mount(window.document.body, {})
        let button = window.document.body.querySelector('button')
        
        button.click()

        expect(ctx.fire).toHaveBeenCalled()
      })

      it('handler should be called with same scope', () => {
        let testClass = class TestClass { foo(){this.bar()} }
        let ctx = new testClass()
        ctx.bar = jest.fn()
        let template = `<button on-click="this.foo" />`
        let compiled = compiler.compile(template, options)
        let mount = eval(compiled).bind(ctx)
        mount(window.document.body, {})
        let button = window.document.body.querySelector('button')

        button.click()

        expect(ctx.bar).toHaveBeenCalled()
      })
    })
  })
})
