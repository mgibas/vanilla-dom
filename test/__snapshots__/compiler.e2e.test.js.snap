// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compiler mounting compiled code arrays mounts array using initial state data - access index 1`] = `"<ul><!--el_1--><li>index: 0</li><li>index: 1</li></ul>"`;

exports[`compiler mounting compiled code arrays mounts array using initial state data - custom "as" 1`] = `"<ul><!--el_1--><li>foo</li><li>bar</li></ul>"`;

exports[`compiler mounting compiled code arrays mounts array using initial state data - custom "index" 1`] = `"<ul><!--el_1--><li>foo</li><li>bar</li></ul>"`;

exports[`compiler mounting compiled code arrays mounts array using initial state data - js expression 1`] = `"<ul><!--el_1--><li>foo</li><li>bar</li><li>foo</li><li>bar</li></ul>"`;

exports[`compiler mounting compiled code arrays mounts array using initial state data - nested template 1`] = `"<ul><!--el_1--><li><p>some text</p><div><span>more</span></div></li><li><p>some text</p><div><span>more</span></div></li></ul>"`;

exports[`compiler mounting compiled code arrays mounts array using initial state data - simple 1`] = `"<ul><!--el_1--><li>foo</li><li>bar</li></ul>"`;

exports[`compiler mounting compiled code arrays mounts array using initial state data - with attributes 1`] = `"<ul><!--el_1--><li><p data-name=\\"foo\\">foo</p></li><li><p data-name=\\"bar\\">bar</p></li></ul>"`;

exports[`compiler mounting compiled code arrays mounts array using initial state data - with sybling 1`] = `"<div><!--el_1--><span>0</span><span>1</span><div>I like this place</div></div>"`;

exports[`compiler mounting compiled code arrays splicing - add item 1`] = `"<ul><!--el_1--><li>foo</li><li>bar</li><li>oof</li></ul>"`;

exports[`compiler mounting compiled code arrays splicing - no array to array with items 1`] = `"<ul><!--el_1--><li>foo</li><li>bar</li></ul>"`;

exports[`compiler mounting compiled code arrays splicing - remove item 1`] = `"<ul><!--el_1--><li>foo</li><li>oof</li></ul>"`;

exports[`compiler mounting compiled code arrays update without array length change - custom "as" 1`] = `"<ul><!--el_1--><li>foo-updated</li><li>bar-updated</li></ul>"`;

exports[`compiler mounting compiled code arrays update without array length change - custom "index" 1`] = `"<ul><!--el_1--><li>foo-updated</li><li>bar-updated</li></ul>"`;

exports[`compiler mounting compiled code arrays update without array length change - js expression 1`] = `"<ul><!--el_1--><li>foo-updated</li><li>bar-updated</li><li>foo-updated</li><li>bar-updated</li></ul>"`;

exports[`compiler mounting compiled code arrays update without array length change - simple 1`] = `"<ul><!--el_1--><li>foo-updated</li><li>bar-updated</li></ul>"`;

exports[`compiler mounting compiled code mount function evaluates given expression - boolean expression false 1`] = `"<div>false</div>"`;

exports[`compiler mounting compiled code mount function evaluates given expression - boolean expression true 1`] = `"<div bar=\\"\\">true</div>"`;

exports[`compiler mounting compiled code mount function evaluates given expression - javascript expressions 1`] = `"<div sum=\\"7\\">5</div>"`;

exports[`compiler mounting compiled code mount function evaluates given expression - method expressions 1`] = `"<div upper=\\"MACIEJ GIBAS\\">MACIEJ</div>"`;

exports[`compiler mounting compiled code mount function evaluates given expression - multiple expressions 1`] = `"<div foo=\\"2 and 3\\">2 and 3</div>"`;

exports[`compiler mounting compiled code mount function evaluates given expression - multiple expressions with undefined properties 1`] = `"<div>  and </div>"`;

exports[`compiler mounting compiled code mount function evaluates given expression - nested property expression on undefined property 1`] = `"<div></div>"`;

exports[`compiler mounting compiled code mount function evaluates given expression - property expression on undefined property 1`] = `"<div></div>"`;

exports[`compiler mounting compiled code mount function evaluates given expression - property expression with 0 as result 1`] = `"<div foo=\\"0\\">0</div>"`;

exports[`compiler mounting compiled code mount function evaluates given expression - property expression with empty string as result 1`] = `"<div></div>"`;

exports[`compiler mounting compiled code mount function evaluates given expression - simple expressions 1`] = `"<div foo=\\"2\\">Maciej Gibas</div>"`;

exports[`compiler mounting compiled code mount function recreates given template on a given node - empty 1`] = `""`;

exports[`compiler mounting compiled code mount function recreates given template on a given node - multiple nodes 1`] = `"<div></div><p></p>"`;

exports[`compiler mounting compiled code mount function recreates given template on a given node - nested 1`] = `"<div><p>hello</p></div>"`;

exports[`compiler mounting compiled code mount function recreates given template on a given node - simple 1`] = `"<div></div>"`;

exports[`compiler mounting compiled code mount function recreates given template on a given node - text nodes 1`] = `
" <div></div>some text and empty line
        
        <p></p>"
`;

exports[`compiler mounting compiled code mount function recreates given template on a given node - with attributes 1`] = `"<div foo=\\"5\\" bar=\\"some\\"></div>"`;

exports[`compiler mounting compiled code update state in dom on update call - boolean expression from empty to non empty 1`] = `"<div bar=\\"blah\\">blah</div>"`;

exports[`compiler mounting compiled code update state in dom on update call - boolean expression from false to true 1`] = `"<div bar=\\"\\">true</div>"`;

exports[`compiler mounting compiled code update state in dom on update call - boolean expression from non empty to empty 1`] = `"<div></div>"`;

exports[`compiler mounting compiled code update state in dom on update call - boolean expression from true to false 1`] = `"<div>false</div>"`;

exports[`compiler mounting compiled code update state in dom on update call - javascript expressions 1`] = `"<div sum=\\"802\\">800</div>"`;

exports[`compiler mounting compiled code update state in dom on update call - method expressions 1`] = `"<div upper=\\"FOO BAR\\">FOO</div>"`;

exports[`compiler mounting compiled code update state in dom on update call - multiple expressions 1`] = `"<div foo=\\"300 and 500\\">300 and 500</div>"`;

exports[`compiler mounting compiled code update state in dom on update call - multiple expressions with undefined properties 1`] = `"<div>  and </div>"`;

exports[`compiler mounting compiled code update state in dom on update call - nested property expression on undefined property 1`] = `"<div></div>"`;

exports[`compiler mounting compiled code update state in dom on update call - property expression from 0 value to non 0 1`] = `"<div foo=\\"1\\">1</div>"`;

exports[`compiler mounting compiled code update state in dom on update call - property expression from non 0 to 0 value 1`] = `"<div foo=\\"0\\">0</div>"`;

exports[`compiler mounting compiled code update state in dom on update call - property expression on undefined property 1`] = `"<div></div>"`;

exports[`compiler mounting compiled code update state in dom on update call - simple expressions 1`] = `"<div foo=\\"300\\">Foo Bar</div>"`;
